crates/sound/src/device.rs

```diff
@@ -469,14 +469,16 @@ impl VhostUserSoundThread {
                         .into());
                     }
                     TxState::WaitingBufferForStreamId(_stream_id) => {
-                        let mut buf = vec![0; descriptor.len() as usize];
-                        let bytes_read = desc_chain
-                            .memory()
-                            .read(&mut buf, descriptor.addr())
-                            .map_err(|_| Error::DescriptorReadFailed)?;
-                        buf.truncate(bytes_read);
-
-                        buffers.push(Buffer::new(buf, Arc::clone(&message)));
+                        /*
+                        Rather than copying the content of a descriptor, buffer keeps a pointer to it.
+                        When we copy just after the request is enqueued, the guest's userspace may or
+                        may not have updated the buffer contents. Guest driver simply moves buffers
+                        from the used ring to the available ring without knowing whether the content
+                        has been updated. The device only reads the buffer from guest memory when the
+                        audio engine requires it. The delay added by this allows the buffer to be
```

Comment:
More than a delay, does this create a kind of synchronization between the producer (driver) and consumer (audio backend) based on the negotiated period?

Reply:
Would it work to replace last two sentences by ``The device only reads the buffer from guest memory when the audio engine requires it, which is after a period thus ensuring that the buffer is up-to-date.``?

Reply:
yep, looks better ;-)

Reply:
Is anyone looking at fixing the driver itself. This seems like a workround for the driver violating the spec.

Reply:
Yep, Matias reported it upstream and they are discussing next steps for the driver and specs: https://lists.oasis-open.org/archives/virtio-comment/202309/msg00175.html

Reply:
This may be a workaround, but I'm not sure. Even if the driver behaves correctly, the device can't consume requests faster than (requests per period) for a given stream. Instead of copying the buffer content, this commit keeps a "pointer" instead, which could also be considered an optimization over keeping a copy of it.

Reply:
I guess the buffer refs ensure we never return a buffer until it has finished being consumed?

Reply:
This commit did not change the way we return buffers. This still happens after a whole period_bytes has been consumed and a MessageIO is dropped.